elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +267,255 @@ private Object fetchRelationship(Environment context) {
+            }
+        }
+        return upsertedObjects;
+    }
+
+    /**
+     * handle upsert on non root-level entities
+     * @param idFieldName id field name
+     * @param input input data object
+     * @param dictionary entity dictionary
+     * @param parentSource parent {@link PersistentResource} object
+     * @param field graphql-java field type
+     * @param outputType graphql-java output type
+     * @return list/set of {@link PersistentResource} objects
+     */
+    private Object upsertObject(String idFieldName, Map<String, Object> input, EntityDictionary dictionary,
+                                PersistentResource parentSource, Field field, GraphQLType outputType,
+                                RequestScope requestScope) {
+        /* fetch id value fed to 'data' argument */
+        Optional<Object> id = input.entrySet().stream()
+                .filter(entry -> idFieldName.equalsIgnoreCase(entry.getKey()))
+                .map(Map.Entry::getValue)
+                .findFirst();
+        Optional<String> idField = Optional.empty();
+        if(id.isPresent()) {
+            idField = Optional.of((String) id.get());
+        }
+
+        Set<Object> fetchEntries;
+        try {
+            fetchEntries = (Set<Object>) fetchObject(dictionary, parentSource, field, outputType.getClass().getName());
+        } catch (BadRequestException e) {
+            log.debug(e.toString());
+            return createObject(input, requestScope, idField, field, parentSource);
+        }
+
+        if(fetchEntries == null) {
+            return createObject(input, requestScope, idField, field, parentSource);
+        } else {
+            return updateObject(fetchEntries, input);
+        }
+    }
+
+    /**
+     * creates or updates a new/existing root-level entity
+     * @param idFieldName field name
+     * @param input input data object
+     * @param requestScope request scope
+     * @param outputType graphql-java output type
+     * @return list/set of {@link PersistentResource} objects
+     */
+    private Object upsertObject(String idFieldName, Map<String, Object> input,
+                                RequestScope requestScope, GraphQLType outputType) {
+        /* fetch id values fed to 'data' argument */
+        Optional<Map.Entry<String, Object>> id = input.entrySet().stream()
+                .filter(entry -> idFieldName.equalsIgnoreCase(entry.getKey()))
+                .findFirst();
+        String idField = null;
+        if(id.isPresent()) {
+            idField = (String) id.get().getValue();
+        }
+
+        Set<Object> fetchEntries;
+        try {
+            fetchEntries = (Set) fetchObject(requestScope, outputType,
+                    Optional.ofNullable(Arrays.asList(idField)), Optional.empty(),
+                    Optional.empty(), Optional.empty(), Optional.empty());
+        } catch(BadRequestException e) { /* no id(s) provided, must create new */
+            log.debug(e.toString());
```

Comment:
You can instead just create the object if the ID is missing in the provided data.